packages/vm/src/wasm_backend/limiting_tunables.rs

```diff
@@ -0,0 +1,219 @@
+use std::ptr::NonNull;
+use std::sync::Arc;
+
+use wasmer::{
+    vm::{self, MemoryError, MemoryStyle, TableStyle, VMMemoryDefinition, VMTableDefinition},
+    MemoryType, Pages, TableType,
+};
+use wasmer_engine::Tunables;
+
+/// A custom tunables that allows you to set a memory limit.
+///
+/// After adjusting the memory limits, it delegates all other logic
+/// to the base tunables.
+pub struct LimitingTunables<T: Tunables> {
+    /// The maxium a linear memory is allowed to be (in Wasm pages, 65 KiB each).
+    /// Since Wasmer ensures there is only none or one memory, this is practically
+    /// an upper limit for the guest memory.
+    limit: Pages,
+    /// The base implementation we delegate all the logic to
+    base: T,
+}
+
+impl<T: Tunables> LimitingTunables<T> {
+    pub fn new(base: T, limit: Pages) -> Self {
+        Self { limit, base }
+    }
+
+    /// Takes in input memory type as requested by the guest and sets
+    /// a maximum if missing. The resulting memory type is final if
+    /// valid. However, this can produce invalid types, such that
+    /// validate_memory must be called before creating the memory.
+    fn adjust_memory(&self, requested: &MemoryType) -> MemoryType {
+        let mut adjusted = *requested;
+        if requested.maximum.is_none() {
+            adjusted.maximum = Some(self.limit);
+        }
+        adjusted
+    }
+
+    /// Ensures the a given memory type does not exceed the memory limit.
+    /// Call this after adjusting the memory.
+    fn validate_memory(&self, ty: &MemoryType) -> Result<(), MemoryError> {
+        if ty.minimum > self.limit {
+            return Err(MemoryError::Generic(
+                "Minimum exceeds the allowed memory limit".to_string(),
+            ));
+        }
+
+        if let Some(max) = ty.maximum {
+            if max > self.limit {
+                return Err(MemoryError::Generic(
+                    "Maximum exceeds the allowed memory limit".to_string(),
+                ));
+            }
+        } else {
+            return Err(MemoryError::Generic("Maximum unset".to_string()));
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: Tunables> Tunables for LimitingTunables<T> {
```

Comment:
I am unclear from here what is called when a new page is allocated in an existing dynamic memory.

Or no need to intercept that, as we ensure valid minimum/maximum when they create the memory table in the first place, so the internal logic can handle dynamic page allocation?

Also, no intercept on pages means we cannot charge gas per memory usage, right? (No need to do so, I just remember the idea was floating around and curious if possible with the current API)

Reply:
Right, we set a maximum once and Wasmer checks that the number of pages does not exceed the maximum internally.

> Also, no intercept on pages means we cannot charge gas per memory usage, right?

Basically yes.

In the new metering we could implement special treatment for the memory grop op code. But then we also need to look at the minimum. So if we really really want, it can be done. But I'd say memory usage if free within the allowed range.